### PR TITLE
feat: Define actions that result in Create vs Edit activities

### PR DESCRIPTION
**Pull request by arvid-e**
https://redmine.weseek.co.jp/issues/177734
https://redmine.weseek.co.jp/issues/177877

## Rules for Update Activity
- https://dev.growi.org/6992e8c71f452b96f288b4cd

### CODE REVIEW

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 1]
```diff
@@ -118,24 +118,69 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
-    const activityEvent = crowi.events.activity;
-    activityEvent.emit(
-      'update',
-      res.locals.activity._id,
-      parameters,
-      { path: updatedPage.path, creator },
-      preNotifyService.generatePreNotify,
-    );
+    const minimumRevisionForActivity = 2;
+    const suppressUpdateWindow = 5 * 60 * 1000; // 5 min
```
Comment (miya) [inline apps/app/src/server/routes/apiv3/page/update-page.ts]: これは定数として route の外に出しおきたいです

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 2]
```diff
@@ -118,24 +118,69 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
+      };
+      const activityEvent = crowi.events.activity;
+      activityEvent.emit(
+        'update',
+        res.locals.activity._id,
+        parameters,
+        { path: updatedPage.path, creator },
+        preNotifyService.generatePreNotify,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (miya) [reply apps/app/src/server/routes/apiv3/page/update-page.ts]: 対応した方がいいと思います

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 3]
```diff
@@ -118,24 +118,69 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
+          SupportedAction.ACTION_PAGE_CREATE,
+          SupportedAction.ACTION_PAGE_UPDATE,
+        ],
+      },
+      _id: { $ne: res.locals.activity?._id },
+    }).sort({ createdAt: -1 });
+
+    // Check if last updated user is me
+    const isLastActivityByMe =
+      lastContentActivity &&
```
Comment (miya) [inline apps/app/src/server/routes/apiv3/page/update-page.ts]: null チェックをしましょう

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 4]
```diff
@@ -118,24 +118,69 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
+      lastContentActivity.user._id.toString() === req.user._id.toString();
```
[the remark this answers is not in the document]
Comment (miya) [reply apps/app/src/server/routes/apiv3/page/update-page.ts]: user が存在しないケースは存在するのか？

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 5]
```diff
@@ -118,24 +118,69 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
+      });
+
+      shouldGenerateUpdateActivity = revisionCount > minimumRevisionForActivity;
+    }
+
+    if (shouldGenerateUpdateActivity) {
+      // persist activity
+      const creator =
+        updatedPage.creator != null
+          ? getIdForRef(updatedPage.creator)
```
[the remark this answers is not in the document]
Comment (miya) [reply apps/app/src/server/routes/apiv3/page/update-page.ts]: `ACTION_UNSETTLED` のアクティビティがそのまま残るのであれば問題なさそう？

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 6]
```diff
@@ -118,24 +118,69 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
-    };
-    const activityEvent = crowi.events.activity;
-    activityEvent.emit(
-      'update',
-      res.locals.activity._id,
-      parameters,
-      { path: updatedPage.path, creator },
-      preNotifyService.generatePreNotify,
-    );
+    const minimumRevisionForActivity = 2;
```
Comment (miya) [inline apps/app/src/server/routes/apiv3/page/update-page.ts]: これは 2 でいいんでしたっけ？

Comment (arvid-e) [reply apps/app/src/server/routes/apiv3/page/update-page.ts]: 2が正しいでした:
- Revision 1: ページ作成
- Revision 2: 初めての編集  (activityが発生しない)
- Revision 3: ２つ目の編集  (activityが発生する)

File: apps/app/src/server/routes/apiv3/page/update-page.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (arvid-e) [reply apps/app/src/server/routes/apiv3/page/update-page.ts]: 前編集したユーザーが削除された場合は`lastContentActivity.user` がnullになるかもしれないので対応しました

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 7]
```diff
@@ -52,6 +52,9 @@ interface UpdatePageRequest
   user: IUserHasId;
 }
 
+const minimumRevisionForActivity = 2;
+const suppressUpdateWindowMs = 5 * 60 * 1000; // 5 min
```
Comment (miya) [inline apps/app/src/server/routes/apiv3/page/update-page.ts]: 定数は大文字且つ snake case にしたいです

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 8]
```diff
@@ -118,24 +121,72 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
+        },
+        _id: { $ne: res.locals.activity?._id },
+      }).sort({ createdAt: -1 });
+
+      const isLastActivityByMe =
+        req.user?._id != null &&
+        lastContentActivity?.user?._id?.toString() ===
+          req.user?._id?.toString();
+
+      const lastActivityTime = lastContentActivity
```
Comment (miya) [inline apps/app/src/server/routes/apiv3/page/update-page.ts]: null チェックをしてください

File: apps/app/src/server/routes/apiv3/page/update-page.ts (hunk 6 above)
Comment (miya) [reply apps/app/src/server/routes/apiv3/page/update-page.ts]: 実装と仕様書 (dev wiki) を一致させてください

File: apps/app/src/server/routes/apiv3/page/update-page.ts
Comment (yuki-takei) [inline apps/app/src/server/routes/apiv3/page/update-page.ts:1]: アクティビティに関する処理がかなり多く入り込むので、別メソッド or 別モジュールにわけて、テストも書いてください

テストは wiki のルール案を満たすようなテストである必要があります
AI でレビューする際は、 skills/learned/essential-test-design を使ってください

File: apps/app/src/server/service/activity/update-activity-logic.ts
Comment (yuki-takei) [inline apps/app/src/server/service/activity/update-activity-logic.ts:1]: Wiki ([/user/arvid-e/Activity Log - Page update rules](https://dev.growi.org/6992e8c71f452b96f288b4cd)) を Opus 4.6 にインプットしたところ、以下のレスポンスが返ってきた

Wiki が正しいとするならば、テストと実装を修正してください

Wiki のルールの方を見直したいならば、口頭で説明にきてください


## PR #10818 Review: `shouldGenerateUpdate` Rule Conformance Check

I've compared the rule specification with the implementation and tests. Below is a summary of findings.

---

### Rule Decision Tree (Organized)

```
1. Is the latest create/update activity on the page mine?
   ├─ YES (isLastActivityByMe)
   │   ├─ Has 5+ minutes passed since that activity?
   │   │   ├─ YES
   │   │   │   ├─ Are there only 2 revisions? (i.e., first update)
   │   │   │   │   ├─ YES → ❌ Do NOT generate update activity
   │   │   │   │   └─ NO  → ✅ Generate update activity
   │   │   └─ NO → ❌ Do NOT generate update activity
   └─ NO → ✅ Generate update activity
```

---

### Implementation Logic (`update-activity-logic.ts`)

```typescript
if (!isLastActivityByMe) {
  shouldGenerateUpdateActivity = true;
} else if (timeSinceLastActivityMs < SUPPRESION_UPDATE_WINDOW_MS) {
  shouldGenerateUpdateActivity = false;
} else {
  const revisionCount = await Revision.countDocuments({ pageId: targetPageId });
  shouldGenerateUpdateActivity = revisionCount > MINIMUM_REVISION_FOR_ACTIVITY; // > 2
}
```

---

### 🔴 Issue 1: Contradiction Between Scenario 4.2 and the Decision Tree

**Decision tree says:**
> Mine → 5+ min elapsed → only 2 revisions (first update) → **Do NOT generate**

**Scenario 4.2 says:**
> Create → come back **30 min later** → first update → "Enough time has passed since create, so don't treat it as part of create" → **Generate update activity**

These two directly contradict each other. **The implementation follows the decision tree** (does NOT generate when revision count is exactly 2, regardless of elapsed time). This means scenario 4.2's expected behavior ("generate") is **not implemented**.

**Action required:** Clarify the intended behavior for "first update after a long gap since create by the same user" and align the decision tree, implementation, and tests accordingly.

---

### 🔴 Issue 2: Duplicate Test Cases

The following two tests have **identical conditions, data, and expected results**:

- **Line 206:** `should generate update activity if: update is by the same user, outside the suppression window, not first update`
- **Line 322:** `should generate update activity if: update is made by the same user, outside suppression window, not first update`

Both insert the same activities (create 2h ago, update 1h ago, current update now), the same 3 revisions, and expect `true`.

**Action required:** Remove the duplicate or modify one to cover a different edge case.

---

### 🟡 Issue 3: Missing Test Coverage for Several Scenarios

| Scenario | Description | Test Present? |
|----------|-------------|:---:|
| **0** | Create only, no update ever made | ❌ |
| **1** | Create → first update within 1 min (same user) | ❌ (line 162 is close but uses 1 hour gap) |
| **3** | Create → update → another update 1 min later (suppression) | ✅ (line 264) |
| **4.1** | Create → come back 1 min later → first update | ❌ |
| **4.2** | Create → come back 30 min later → first update | ❌ |
| **5.1 / 5.2** | Edits interleaved with other page edits (changed to "don't care about other pages") | ❌ (implementation correctly scopes to the same page, but no explicit test) |

**Action required:** Add tests for the missing scenarios to ensure comprehensive coverage.

---

### ✅ What's Correct

| Rule Path | Implementation | Status |
|-----------|---------------|:---:|
| Last activity is NOT mine → generate | `!isLastActivityByMe → true` | ✅ |
| Last activity is mine + within 5 min → suppress | `timeSinceLastActivityMs < 5min → false` | ✅ |
| Last activity is mine + 5+ min + revisions > 2 → generate | `revisionCount > 2 → true` | ✅ |
| Last activity is mine + 5+ min + revisions ≤ 2 → suppress | `revisionCount <= 2 → false` | ✅ |

---

### 📝 Summary of Required Actions

| Priority | Item | Action |
|----------|------|--------|
| 🔴 Critical | Scenario 4.2 vs. decision tree contradiction | Decide intended behavior, then fix either the spec or the implementation |
| 🔴 High | Duplicate tests (lines 206 & 322) | Remove one or differentiate them |
| 🟡 Medium | Missing test coverage (scenarios 0, 1, 4.1, 4.2, 5.x) | Add tests for these edge cases |

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 9]
```diff
@@ -118,24 +119,41 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
-      res.locals.activity._id,
-      parameters,
-      { path: updatedPage.path, creator },
-      preNotifyService.generatePreNotify,
-    );
+    // Decide if update activity should generate
+    try {
+      const targetPageId = updatedPage._id.toString();
+      const currentActivityId = res.locals.activity?._id.toString();
+      const currentUserId = req.user?._id?.toString();
```
Comment (yuki-takei) [inline apps/app/src/server/routes/apiv3/page/update-page.ts]: currentUserId が undefined の場合、isLastActivityByMe は常に false となり、常にアクティビティが生成されてしまう。認証済みルートなので実際には起きにくいが、型定義を string | undefined に明示し、undefined時の振る舞いを意図的にすべき。

File: apps/app/src/server/service/activity/update-activity-logic.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        SupportedAction.ACTION_PAGE_CREATE,
+        SupportedAction.ACTION_PAGE_UPDATE,
+      ],
+    },
+    _id: { $ne: currentActivityId },
+  }).sort({ createdAt: -1 });
+
+  const isLastActivityByMe =
+    !!currentUserId &&
+    lastContentActivity?.user?._id?.toString() === currentUserId;
```
Comment (yuki-takei) [inline apps/app/src/server/service/activity/update-activity-logic.ts]: `import { getIdForRef, getIdStringForRef } from '@growi/core';`

を利用してください

File: apps/app/src/server/service/activity/update-activity.spec.ts [hunk 11]
```diff
@@ -0,0 +1,460 @@
+import { MongoMemoryServer } from 'mongodb-memory-server-core';
+import mongoose from 'mongoose';
+
+import { SupportedAction } from '~/interfaces/activity';
+import Activity from '~/server/models/activity';
+import { Revision } from '~/server/models/revision';
```
Comment (yuki-takei) [inline apps/app/src/server/service/activity/update-activity.spec.ts:6]: モデルの import 方法は実装とテストで揃えて

File: apps/app/src/server/routes/apiv3/page/update-page.ts [hunk 12]
```diff
@@ -118,24 +119,41 @@ export const updatePageHandlersFactory = (crowi: Crowi): RequestHandler[] => {
...
+          'update',
+          res.locals.activity._id,
+          parameters,
+          { path: updatedPage.path, creator },
+          preNotifyService.generatePreNotify,
+        );
+      }
+    } catch (err) {
+      logger.error('Failed to generate update activity', err);
+    }
```
Comment (yuki-takei) [inline apps/app/src/server/routes/apiv3/page/update-page.ts:164]: try-catch をいくつかの処理ブロックに分割してください

**Review by yuki-takei — CHANGES_REQUESTED** [review]
